core/src/storage2/collections/bitstash/counts.rs

```diff
@@ -63,24 +63,57 @@ impl FullMask {
     pub fn reset_full(&mut self, index: u8) {
         self.0 &= !(1_u32 << (31 - index as u32));
     }
+
+    /// Returns `true` if there is no more space available at any of the indices
+    /// and hence all chunks are full.
+    pub fn is_completely_full(self) -> bool {
+        self.0 == u32::MAX
+    }
 }
 
 impl CountFree {
     /// Returns the position of the first free `u8` in the free counts.
     ///
     /// Returns `None` if all counts are `0xFF`.
     pub fn position_first_zero(&mut self) -> Option<u8> {
-        for (i, count) in self.counts.iter_mut().enumerate() {
-            if !self.full.is_full(i as u8) {
-                if *count == !0 {
-                    self.full.set_full(i as u8);
-                } else {
-                    *count += 1;
-                }
-                return Some(i as u8)
-            }
+        if self.full.is_completely_full() {
+            return None
+        }
+
+        let i = (!self.full.0).leading_zeros();
+        if i == 32 {
+            unreachable!(
+                "if full we must already have exited right at the start of this function"
+            );
+        }
+
+        Some(i as u8)
```

Comment:
Have you experimented around with removing `self.full.is_completely_full()` and instead make `unreachable!` just return `None`? Would look a bit clearer and in the common case we drop one integer comparison.